### ISSUE TITLE
Empower providers to provide advisories in different CSAF versions in parallel

### ISSUE THREAD

**Issue by sthagen** [oasis-tcs/csaf#843]
During the CSAF Community Days 2024, 12.-13. December 2024 in Munich, it became clear, that providers often map directories directly served per HTTPS to SCAF format versions.

> Examples: `foo/v2/bar` or `baz/2.0/quux`.

>Other providers do not indicate any CSAF format version in the paths or the provider.json
resource.

All these use cases are plausible and fit the provider and consumer needs it seems,
but when CSAF v2.1 will be available there is the reasonable assumption, that for some
time out of control of the TC there will be many mixed format situations:

- Only CSAF v2.0
- CSAF v2.0 for older advisories and CSAF v2.1 for "fresher" advisories
- CSAF v2.0 for all advisories and CSAF v2.1 for "fresher" advisories
- Only CAF v2.1

Questions to seed the discussion:

- Do we need to adjust the provider json schema of CSAF v2.1?
- Do we need to amend also the CSAF v2.0 provider json schema?
- Do we lack informative text in our CSAF v2.1 prose to guide and support the communities?
- Would the CSAF eco system profit from FAQ entries providing solutions for these scenarios?

**Comment by tschmidtb51** [comment]
Here is the intended process:

- The PMD that should be used resides at `.well-known/csaf/provider-metadata.json`. It is actually a symlink to one of the PMDs (either v2.0 or v2.1).
- The actual PMD and related files are locate in a specific folder:
  - `.well-known/csaf/v2.0/` for CSAF 2.0
  - `.well-known/csaf/v2.1/` for CSAF 2.1
- This allows to serve 2 versions in parallel.
- *Note: The PMDs need to use their true canonical URL to allow to be pinned.`

The transition would look like this:
- Setup the folder structure; set the symlink to use the CSAF 2.0 PMD.
- Begin to serve CSAF 2.1 (as non-active version) for time frame a (e.g. as beta)
- Announce change to CSAF 2.1 with fixed date (and announce how long the CSAF 2.0 data will be maintained, might be 0h)
- convert all CSAF 2.0 to CSAF 2.1 data
- Change symlink to serve CSAF 2.1 PMD
- Still maintain CSAF 2.0 for the announce time frame
- Stop to maintain CSAF 2.0
- Stop to serve CSAF 2.0
- optional: move CSAF 2.0 files as archive to `.well-known/csaf/archive/v2.0.tar.bz`

**Comment by tschmidtb51** [comment]
Todos:

- [ ] add `maintained_until` to PMD
- [ ] add `maintained_from` to PMD
- [ ] add requirement to warn if `maintained_until` in PMD is less then 90 days
- [ ] add requirement to error if `maintained_until` in PMD has passed => supersedable with `--force-use-old-version` option
- [ ] add transition description
- [ ] add requirement in CSAF management for user option to prefer one version (default: use current)
- [ ] add fields to `aggregator.json`
- [ ] clarify that aggregators MUST NOT convert documents to older/new versions => just multipliers are allowed to do so; requirement to add disclaimer text
- [ ] check whether additional test are needed in section 7
- [ ] clarify that CSAF downloader needs to work with direct PMD

**Comment by oxisto** [comment]
Having given this some though, this is really not an easy situation. Personally, I am not convinced that there is much to gain my serving two different versions of the provider metadata. An ideal situation would be that if a provider wants to support CSAF 2.1 we would immediately update its PMD to 2.1 and then
- either immediately convert all 2.0 documents to 2.1, with the feed staying the same or
- go into some transition phase where the existing 2.0 documents are served at the existing feed / directly (until a given time) and new documents will be served as an extra feed

The obvious problem here is that a 2.1 PMD will not be able to be parsed by a fully compliant 2.0 library. I wonder if we could go back to @sthagen's original idea of amending the 2.0 standard, especially when it comes to the `provider_metadata` field. Unfortunately this is a fixed enum of `2.0` in 2.0 and I wonder whether a regex allowing also future versions would be a better solution - similar to what CVSS is/was doing with 3.0/3.1

**Comment by tschmidtb51** [comment]
@oxisto The extra fields can obviously only be introduced in (and used) from CSAF 2.1 onward.

Serving two versions in parallel is a common transition use case: Not all customers / consumers will instantly be able to parse CSAF 2.1. So you provide for a time x both, CSAF 2.0 and CSAF 2.1. The consumer choose which one should be preferred. The transition starts with adding CSAF 2.1 (e.g. as beta; maybe, you also want to test your CSAF 2.1 implementation before you rely on it in production) while still keeping CSAF 2.0 as default. Then, you add CSAF 2.1 into your `security.txt` and thus make it accessible. After a given period of time, you change the default to CSAF 2.1 while keeping and maintaining CSAF 2.0. A little later, you drop the support for CSAF 2.0
BTW: The same transition was done during the change from CVRF to CSAF.

**Comment by oxisto** [comment]
Just to clarify: Who should warn or throw the error if the `maintained_until` is inappropriate ? Would that be a requirement for a publisher "server" or a user? I am currently adding this to requirement 7 in a rather generic way since we do not really have "tests" for either a distribution "server" or consumer.

**Comment by oxisto** [comment]
As agreed with @tschmidtb51 `aggregator.json` handling is split off to #992

**Comment by tschmidtb51** [comment]
> optional: move CSAF 2.0 files as archive to `.well-known/csaf/archive/v2.0.tar.bz`

@sthagen: The article that indicates `xz` would be a bad choice: https://www.nongnu.org/lzip/xz_inadequate.html

**Comment by sthagen** [comment]
Thanks for the interesting link. I would still prefer `xz` as compression, because these days we do not rely on bit correction of unique artifacts, but instead prefer well-hashed space-efficientrly compressed artifacts, where we select from copies per this integroity measure.

I am OK with the optional `bzip2` recommendation, but suggest to add a `2` in:

> optional: move CSAF 2.0 files as archive to `.well-known/csaf/archive/v2.0.tar.bz`

I think it should be (`tar -cJf <some-fs-place>/.well-known/csaf/archive/v2.0.tar.bz2 <file-tree>`):

> optional: move CSAF 2.0 files as archive to `.well-known/csaf/archive/v2.0.tar.bz2`

As we do want to avoid the original bzip program, right?

From my local tar man page:

>     -j, --bzip2
>              Filter the archive through bzip2(1).

**Comment by tschmidtb51** [comment]
> As we do want to avoid the original bzip program, right?

Definitely. I add the the `2`. Thanks for spotting that.

**Comment by tschmidtb51** [comment]
@oxisto: In that case, we should at least require for the hashes to be added to the archive. Also, it should be suggested to put them in year folder based on their `initial_release_date`.

**Comment by bernhardreiter** [comment]
> optional: move CSAF 2.0 files as archive `.well-known/csaf/archive/v2.0.tar.bz2`

This is a question about what should happen **after** the time frame when two distribution versions of CSAF documents are served in parallel. I  therefore suggest to discuss the pros and cons first, to get to a decision and to have this discussion in a separate issue.

> I would still prefer xz as compression

The choice of the compression algorithm could be delayed after the previous question has been decided. In any case should the discussion go into a new issue. As it is not about a general recommendation about which compression algorithms and implementations are best overall, but what is best for CSAF archiving. And the goals of the archiving are still to be fully determined.

(To give examples of details for the compression question: Availability of implementations on different platforms and programming environments. Newer algorithms like Brotli or zstandard. Keeping old URLs as compression already happens on the HTTP layer.)
